### ISSUE TITLE
[Bug] Adam optimizer produces near-zero effective updates when training LoRA weights in bfloat16 with ModelParallel on TPU

### ISSUE THREAD

**Issue by 14790897** [keras-team/keras-hub#2629]
**Describe the bug**

When fine-tuning a large model (Qwen2.5-7B) with LoRA using `ModelParallel` on TPU v5e-8, 
the Adam/AdamW optimizer produces no effective weight updates: the loss is identical 
at step 1 of every epoch, indicating the LoRA weights are never modified.
Root cause may be: LoRA variables inherit the layer's bfloat16 dtype, and Adam's second-moment estimate v_t underflows in bfloat16, causing near-zero effective step sizes.

**To Reproduce**

keras version: 3.13.2
Use adam: https://www.kaggle.com/code/liuweiq/keras3-jax-qwen2-5-tpu?scriptVersionId=302386192
It's worth noting that Switching to SGD avoids this issue.
Use sgd:  https://www.kaggle.com/code/liuweiq/keras3-jax-qwen2-5-tpu/notebook?scriptVersionId=302428315

**Expected behavior**

loss should change when using adam.
**Additional context**

epoch14(loss: 0.5014):
<img width="1888" height="1048" alt="Image" src="https://github.com/user-attachments/assets/8aed5e52-2ebb-47f2-a4f3-cff7cd218f85" />
epoch15(loss: 0.5014):
<img width="1795" height="1134" alt="Image" src="https://github.com/user-attachments/assets/4b7f769a-4db9-4028-a9dc-35b60a0498b4" />

**Would you like to help us fix it?**
no

**Comment by divyashreepathihalli** [comment]
is this working for you with other optimizers?

**Comment by 14790897** [comment]
Yes, it works with SGD.

> It's worth noting that Switching to SGD avoids this issue.
> Use sgd: https://www.kaggle.com/code/liuweiq/keras3-jax-qwen2-5-tpu/notebook?scriptVersionId=302428315

**Comment by james77777778** [comment]
I've identified a potential issue: we should use float32 for LoRA weights to stabilize finetuning and avoid underflow or overflow.

PR here: https://github.com/keras-team/keras/pull/22559

**Comment by JiwaniZakir** [comment]
This is a classic bfloat16 numerical stability issue with Adam's second-moment accumulator. When LoRA variables are created with `dtype=bfloat16` (inherited from the layer's compute dtype), Adam's optimizer state variables `v_t` are also created in bfloat16 — and since LoRA gradients are small at initialization (lora_B is zero-initialized), the squared gradient values underflow to zero in bfloat16's limited 7-bit mantissa precision, keeping `v_t ≈ 0` and causing `lr * m_t / (sqrt(v_t) + epsilon)` to round to zero when applied back to the bfloat16 weight. SGD avoids this because it has no second-moment term. The fix likely belongs in the LoRA variable creation path — LoRA variables (and/or their optimizer slots) should be forced to float32 dtype regardless of the layer's compute dtype, similar to how Keras's mixed-precision `LossScaleOptimizer` keeps slot variables in float32; the relevant area to look at would be `keras_hub/src/layers/lora_dense.py` and how it sets `dtype` when calling `self.add_weight` for `lora_kernel_a` and `lora_kernel_b`.

**Comment by ManojPawar2** [comment]
I looked into this and was able to pin down the root cause.

  When LoRA is enabled, the adapter weights inherit the layer's `bfloat16` dtype. As a result, Adam accumulates its second-moment
  estimate `v_t` in bfloat16 as well. Because bfloat16 has very limited mantissa precision, the small squared gradients underflow
  toward zero, the adaptive step size collapses, and the weights effectively stop updating. This also explains why SGD trains
  normally while Adam stalls: SGD has no second-moment term, so it is not affected by the underflow.

  Proposed fix: keep the LoRA kernels (`lora_kernel_a` / `lora_kernel_b`) in float32 so the optimizer states are maintained in full
  precision, and cast them down to the compute dtype only within the forward pass. Since the LoRA adapters are very small relative to
  the frozen base weights, the additional memory is negligible, while fine-tuning becomes numerically stable. The change is applied
  across the attention `EinsumDense` projections shared by the affected backbones (Gemma, Qwen, Llama, Mistral, and others).

  The fix is covered by the Gemma, Gemma3, and Gemma4 LoRA fine-tuning and save/reload test suites, all of which pass.

  PR:  **keras-team/keras-hub#2772**. Reviews and feedback are welcome.
